service/raft/group0_state_machine.cc

```diff
@@ -174,7 +196,9 @@ future<> group0_state_machine::merge_and_apply(group0_state_machine_merger& merg
         co_return;
     },
     [&] (write_mutations& muts) -> future<> {
```

Comment:
Here is the specification for `write_mutations` written at the point of definition:
```cpp
// This command is used to write data to tables other than topology or
// schema tables and it doesn't update any in-memory data structures.
struct write_mutations {
    std::vector<canonical_mutation> mutations;
};
```

this new code contradicts that specification. (**and it doesn't update any in-memory data structures**)

Reply:
I think I recall specifically requesting in some discussion, slack or some design doc thread, not to touch `write_mutations` (get off my lawn :P) but create a separate command type

Reply:
Previously this PR added a `service_level_change`, but recently @nuivall's PR added `group0_batch` which collects mutations across whole request (e.g. some commands do changes to both auth and service levels and we want to make this change atomically) and post it as `write_mutation`, so I've dropped the new command type.

Another thing is in https://github.com/scylladb/scylladb/pull/19085 (follow up to this PR) I need to update service levels cache on changes to some auth tables but on only to 2 of them and auth has more. 
So I did a commit similar to this and @gleb-cloudius (https://github.com/scylladb/scylladb/pull/19085#discussion_r1629440154) wrote he thought of sth similar, that's why I did it also here.

Some time ago @piodul proposed to introduce
```
struct {
	enum component_type : uint16_t {
		schema,
		topology,
		// In the near future:
		service_levels,
		auth,
	};

	std::vector<component_type> components_to_update;
	std::vector<canonical_mutation> mutations;
};
```
(https://scylladb-corporate.slack.com/archives/CK7BG6Y6L/p1716557376467699) but AFAIR there was no conclusion or there were some arguments against it.

Would it be okay to create a new command type like `notification_change` and do the same thing as this commit is doing with `write_mutations` (browse table_ids of mutations and update defined subsystems)? 
This way `group0_batch` can return a single raft command and I can still "subscribe" to appropriate tables.

Reply:
> Would it be okay to create a new command type like notification_change and do the same thing as this commit is doing with write_mutations (browse table_ids of mutations and update defined subsystems)?

Yes

IIUC this is something that we agreed on to do eventually -- but the scope of that is larger, we want all components to be update-able correctly that way, but that requires refactoring the components so we can apply RCU style updates to them and update all of them at once as a single isolated transaction.

I think @nuivall agreed to push this effort (in exchange for merging his PR :P)

For now I think we can just create the command type and update only service levels, and later extend the logic to be able to update all components. We just need to make sure that we don't corner ourselves by implementing something that won't be extendable later.

@piodul @nuivall did someone take notes / write a design doc based on the discussions we had and decisions we made?

Let's make sure that introducing `notification_change` is compatible with this so `notification_change` can be later extended with whatever @nuivall wants to do

Reply:
yes, I am trying to push it, making some notes for myself, later maybe transfer to the design doc depending on the need.

from my pov do whatever works here and is inline with current code, likely I will need to refactor it anyway, don't see how it can be forward compatible with something which doesn't yet exists.

Reply:
> from my pov do whatever works here and is inline with current code, likely I will need to refactor it anyway, 

It would be nice if we were able to avoid adding yet another command type later when you have your final design.

On a glance the `notification_change` proposed here by @Jadw1 is similar to what you want to do, so I hope it will be forward-compatible.

Alternatively if this PR is not a priority (and I don't think it is?) we can simply delay merging it until the design is ready.

Reply:
@kbr-scylla this is an important follow-up for service levels on raft, other work depend on it (https://github.com/scylladb/scylladb/pull/19085). It would delay our current effort of fixing bugs in SLs.

Reply:
Well anyway I'm fine with introducing `notification_change` as described by @Jadw1 . @gleb-cloudius what's your opinion?

Reply:
It's a change that causes notifications, so I think "notification change" (short of: a change of the notification kind) is the appropriate one

Alternatively, `notifying_change`

Reply:
Sorry for bikeshedding, but I think the suggested name is confusing. A notification presumes there is a subscriber. The in-memory state of group0 state machine is not a subscriber, it's the object that is being changed. The name "service level change" was good enough.

Reply:
Let me join the bikeshedding. I do not like either name. They sound like we want to add yet another command type while we want to replace all command types by this new one.  I am also not sure why do we need a new command type instead of extending `write_mutations` like the patch does. This is not external API. We can review the current usage and see if it is safe to do so.

Reply:
> Let me join the bikeshedding. I do not like either name. They sound like we want to add yet another command type while we want to replace all command types by this new one. I am also not sure why do we need a new command type instead of extending `write_mutations` like the patch does. This is not external API. We can review the current usage and see if it is safe to do so.

@kbr-scylla's objection to just extending `write_mutations` was that it would violate the current contract (i.e. it doesn't update any in-memory structures). I'm personally fine with changing the contract to allow the current change as we are ultimately going to get rid of `write_mutations` anyway. @kbr-scylla WDYT?

Reply:
**If** you verify that the current usages of `write_mutations` won't be affected.

For example the usage in sending CDC generation data. I don't want those commands to cause any reload of topology state or anything. They should be pure mutation writes.
(It's fine to execute the logic to calculate what needs to be reloaded -- if the result is that nothing needs to be reloaded)

> I'm personally fine with changing the contract to allow the current change as we are ultimately going to get rid of write_mutations anyway. @kbr-scylla WDYT?

Get rid of it, how? You just proposed to extend it so basically going into the direction of making `write_mutations` "the" command responsible for everything in the future.

Reply:
> **If** you verify that the current usages of `write_mutations` won't be affected.
> 
> For example the usage in sending CDC generation data. I don't want those commands to cause any reload of topology state or anything. They should be pure mutation writes. (It's fine to execute the logic to calculate what needs to be reloaded -- if the result is that nothing needs to be reloaded)

In that case I believe they won't affect current use because we only reload service levels cache when there are mutations that modify the service levels table. @Jadw1 please confirm.

> Get rid of it, how? You just proposed to extend it so basically going into the direction of making `write_mutations` "the" command responsible for everything in the future.

Ok, I didn't phrase what I wanted to say well. I meant that we plan to clean up the design of the commands later. I have no idea or no comment on whether `write_mutations` will be reused, I initially thought that we will rather be introducing new commands and deprecating the old ones, but it's probably an unimportant detail from the perspective of now.

Reply:
> In that case I believe they won't affect current use because we only reload service levels cache when there are mutations that modify the service levels table. @Jadw1 please confirm.

That's right. I'm not going to move topology or schema changes to `write_mutations`.

Reply:
To sum up the conversation, I'm following the solution with looking at table_ids in `write_mutations` and I'm not adding any new command type for service levels.